Stratis.Bitcoin/BlockPulling/BlockPuller.cs

```diff
@@ -267,7 +267,7 @@ private void AssignPendingVectors()
 			}
 
 			var minheight = int.MaxValue;
-			foreach (InventoryVector vector in vectors)
+			foreach (InventoryVector vector in vectors.ToArray())
```

Comment:
This is the only functional change. Rest is refactoring.
See individual commits.